### PR TITLE
sql executor for AddAssetQuantity command

### PR DESCRIPTION
**Pull request by victordrobny**
Signed-off-by: Victor Drobny <vdrobny@yandex.ru>

### Description of the Change
AddAssetQuantity command is now executed in sql.

### Benefits
Less db communication, better performance

### CODE REVIEW

File: irohad/ametsuchi/impl/postgres_wsv_command.cpp [hunk 1]
```diff
@@ -404,5 +404,78 @@ namespace iroha {
...
+        const shared_model::interface::types::AssetIdType &asset_id,
+        const std::string &amount,
+        const int precision) {
+      std::string query =
+          (boost::format(
+              // clang-format off
+              R"(
+          WITH has_account AS (SELECT account_id FROM account WHERE account_id = '%s' LIMIT 1),
+               has_asset AS (SELECT asset_id FROM asset WHERE asset_id = '%s' AND precision = %d LIMIT 1),
+               amount AS (SELECT amount FROM account_has_asset WHERE asset_id = '%s' LIMIT 1),
```
Comment (lebdron) [inline irohad/ametsuchi/impl/postgres_wsv_command.cpp]: `SELECT amount FROM account_has_asset WHERE asset_id = '%s' LIMIT 1` will return the first account from the table which has the specified asset. It seems like constraint or account id is missing.

File: irohad/ametsuchi/impl/storage_impl.cpp [hunk 2]
```diff
@@ -181,6 +181,7 @@ DROP TABLE IF EXISTS index_by_id_height_asset;
       pqxx::lazyconnection temp_connection(options_str_without_dbname);
       auto transaction =
           std::make_unique<pqxx::nontransaction>(temp_connection);
+      std::cout << "DB: " << dbname << std::endl;
```
Comment (l4l) [inline irohad/ametsuchi/impl/storage_impl.cpp]: Either remove or use log

File: irohad/ametsuchi/impl/postgres_wsv_command.cpp [hunk 3]
```diff
@@ -404,5 +404,78 @@ namespace iroha {
...
+          END AS result;)"
+           // clang-format on
+          )
+           % account_id % asset_id % precision % asset_id % amount % account_id
+           % asset_id)
+              .str();
+
+      auto result = execute_(query);
+
+      return result.match(
```
Comment (nickaleks) [inline irohad/ametsuchi/impl/postgres_wsv_command.cpp]: Its beautiful

File: irohad/execution/impl/command_executor.cpp [hunk 4]
```diff
@@ -63,90 +63,11 @@ namespace iroha {
...
-          return makeCommandResult(
-              commands->upsertAccountAsset(*account_asset_new_val.value),
-              command_name);
-        },
-        [&command_name](const auto &account_asset_error) -> CommandResult {
-          return makeCommandError("account asset builder failed. reason "
-                                      + account_asset_error.error.toString(),
-                                  command_name);
-        });
+    auto result = commands->addAssetQuantity(creator_account_id,
```
Comment (nickaleks) [inline irohad/execution/impl/command_executor.cpp]: This class seems redundant now. Maybe create Postgres command executor and remove wsv_command altogether?

Comment (nickaleks) [reply irohad/execution/impl/command_executor.cpp]: We can make postgres_command executor and keep not-rewritten functions along with changed ones together until migration is complete

File: test/module/irohad/ametsuchi/wsv_query_command_test.cpp [hunk 5]
```diff
@@ -481,5 +482,99 @@ namespace iroha {
...
+      account_asset =
+          query->getAccountAsset(account->accountId(), asset_id);
+      ASSERT_TRUE(account_asset);
+      ASSERT_EQ("2", account_asset.get()->balance().toStringRepr());
+    }
+
+    /**
+     * @given WSV command
+     * @when trying to add account asset with non-existing asset
+     * @then account asset fails to added
```
Comment (nickaleks) [inline test/module/irohad/ametsuchi/wsv_query_command_test.cpp]: fails to be added

File: irohad/ametsuchi/command_executor.hpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+      /**
+     * AddAssetQuantity sql executor
+     * @return WsvCommandResult, which will contain error in case of failure
+     */
+      virtual WsvCommandResult addAssetQuantity(
+          const shared_model::interface::types::AccountIdType &account_id,
+          const shared_model::interface::types::AssetIdType &asset_id,
+          const std::string &amount,
+          const int precision) = 0;
```
Comment (nickaleks) [inline irohad/ametsuchi/command_executor.hpp]: Precision type?

File: irohad/ametsuchi/impl/mutable_storage_impl.cpp [hunk 7]
```diff
@@ -41,8 +42,9 @@ namespace iroha {
           log_(logger::log("MutableStorage")) {
       auto query = std::make_shared<PostgresWsvQuery>(*transaction_);
       auto command = std::make_shared<PostgresWsvCommand>(*transaction_);
+      auto command_executor = std::make_shared<PostgresCommandExecutor>(*transaction_);
       command_executor_ =
-          std::make_shared<CommandExecutor>(CommandExecutor(query, command));
+          std::make_shared<iroha::CommandExecutor>(iroha::CommandExecutor(query, command, command_executor));
```
Comment (nickaleks) [inline irohad/ametsuchi/impl/mutable_storage_impl.cpp]: Is construction of iroha::CommandExecutor needed here?

File: irohad/ametsuchi/impl/postgres_command_executor.cpp [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+                               // clang-format on
+                               )
+                           % account_id % asset_id % precision % asset_id
+                           % account_id % amount % account_id % asset_id)
+                              .str();
+
+      auto result = execute_(query);
+
+      return result.match(
+          [](expected::Value<pqxx::result> error_code) -> WsvCommandResult {
```
Comment (nickaleks) [inline irohad/ametsuchi/impl/postgres_command_executor.cpp]: Pass by reference?

File: test/module/irohad/ametsuchi/wsv_query_command_test.cpp [hunk 9]
```diff
@@ -481,5 +485,99 @@ namespace iroha {
...
+      void SetUp() override {
+        WsvQueryCommandTest::SetUp();
+
+        ASSERT_TRUE(val(command->insertRole(role)));
+        ASSERT_TRUE(val(command->insertDomain(*domain)));
+        ASSERT_TRUE(val(command->insertAccount(*account)));
+      }
+
+     public:
+      void addAsset() {
```
Comment (nickaleks) [inline test/module/irohad/ametsuchi/wsv_query_command_test.cpp]: Can you add comment about what this function does please

File: irohad/ametsuchi/command_executor.hpp [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+     *  If something goes wrong, Result will contain WsvError
+     *  with additional information
+     */
+    using WsvCommandResult = expected::Result<void, WsvError>;
+
+    class CommandExecutor {
+     public:
+      virtual ~CommandExecutor() = default;
+
+      /**
```
Comment (nickaleks) [inline irohad/ametsuchi/command_executor.hpp]: Format comment please

File: irohad/ametsuchi/impl/postgres_command_executor.cpp [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        const shared_model::interface::types::PrecisionType precision) {
+      std::string query = (boost::format(
+                               // clang-format off
+          R"(
+          WITH has_account AS (SELECT account_id FROM account WHERE account_id = '%s' LIMIT 1),
+               has_asset AS (SELECT asset_id FROM asset WHERE asset_id = '%s' AND precision = %d LIMIT 1),
+               amount AS (SELECT amount FROM account_has_asset WHERE asset_id = '%s' AND account_id = '%s' LIMIT 1),
+               new_value AS (SELECT %s +
+                              (SELECT
+                                  CASE WHEN EXISTS (SELECT amount FROM amount LIMIT 1) THEN (SELECT amount FROM amount LIMIT 1)
```
Comment (nickaleks) [inline irohad/ametsuchi/impl/postgres_command_executor.cpp]: Can we reformat it so it fits in 80 characters?

**Comment by victordrobny** [comment]
All sql executors are in https://github.com/hyperledger/iroha/pull/1557

**Pull request closed without merging**